samza-core/src/main/java/org/apache/samza/operators/spec/InputOperatorSpec.java

```diff
@@ -18,35 +18,46 @@
  */
 package org.apache.samza.operators.spec;
 
-import org.apache.commons.lang3.tuple.Pair;
+import org.apache.samza.operators.KV;
+import org.apache.samza.serializers.Serde;
 import org.apache.samza.system.StreamSpec;
 
-import java.util.function.BiFunction;
-
 /**
  * The spec for an operator that receives incoming messages from an input stream
  * and converts them to the input message.
  *
- * @param <K> the type of key in the incoming message
- * @param <V> the type of message in the incoming message
- * @param <M> the type of input message
+ * @param <K> the type of input key
+ * @param <V> the type of input value
  */
-public class InputOperatorSpec<K, V, M> extends OperatorSpec<Pair<K, V>, M> {
+public class InputOperatorSpec<K, V> extends OperatorSpec<KV<K, V>, Object> { // Object == KV<K, V> | V
 
   private final StreamSpec streamSpec;
-  private final BiFunction<K, V, M> msgBuilder;
+  private final Serde<K> keySerde;
+  private final Serde<V> valueSerde;
+  private final boolean isKeyedInput;
 
-  public InputOperatorSpec(StreamSpec streamSpec, BiFunction<K, V, M> msgBuilder, int opId) {
+  public InputOperatorSpec(StreamSpec streamSpec,
```

Comment:
We talked offline. There's not much meaningful type checking we can do between spec and impl DAGs in the current implementation since the upper bound for message types in specs is Object. The best we can do for now is make the types _within_ the spec and impl DAG consistent. We'll revisit when we have the StreamDescriptor/StreamApplication API.